src/dune_pkg/opam_solver.ml

```diff
@@ -0,0 +1,190 @@
+open Stdune
+
+module type CONTEXT = Opam_0install.S.CONTEXT
+
+(* Helper functor which implements [CONTEXT] for [(L.t, R.t) Either.t] where
+   [L] and [R] both implement [CONTEXT] *)
+module Context_either (L : CONTEXT) (R : CONTEXT) :
+  CONTEXT with type t = (L.t, R.t) Either.t = struct
+  type t = (L.t, R.t) Either.t
+
+  type rejection = (L.rejection, R.rejection) Either.t
+
+  let pp_rejection f = function
+    | Left l -> L.pp_rejection f l
+    | Right r -> R.pp_rejection f r
+
+  let candidates t name =
+    let convert_rejections ~f =
+      List.map ~f:(fun (version, result) ->
+          (version, Result.map_error ~f result))
+    in
+    match t with
+    | Left l -> L.candidates l name |> convert_rejections ~f:Either.left
+    | Right r -> R.candidates r name |> convert_rejections ~f:Either.right
+
+  let user_restrictions = function
+    | Left l -> L.user_restrictions l
+    | Right r -> R.user_restrictions r
+
+  let filter_deps = function
+    | Left l -> L.filter_deps l
+    | Right r -> R.filter_deps r
+end
+
+(* Helper functor which adds a set of local packages to a [CONTEXT] *)
+module Context_with_local_packages (Base_context : CONTEXT) : sig
+  include CONTEXT
+
+  val create :
+       base_context:Base_context.t
+    -> local_packages:OpamFile.OPAM.t OpamPackage.Name.Map.t
+    -> t
+
+  val local_packages : t -> OpamFile.OPAM.t OpamPackage.Name.Map.t
+end = struct
+  let local_package_default_version = OpamPackage.Version.of_string "LOCAL"
+
+  type t =
+    { base_context : Base_context.t
+    ; local_packages : OpamFile.OPAM.t OpamPackage.Name.Map.t
+    }
+
+  let create ~base_context ~local_packages = { base_context; local_packages }
+
+  let local_packages { local_packages; _ } = local_packages
+
+  type rejection = Base_context.rejection
+
+  let pp_rejection = Base_context.pp_rejection
+
+  let candidates t name =
+    match OpamPackage.Name.Map.find_opt name t.local_packages with
+    | None -> Base_context.candidates t.base_context name
+    | Some opam_file ->
+      let version =
+        Option.value opam_file.version ~default:local_package_default_version
+      in
+      [ (version, Ok opam_file) ]
+
+  let user_restrictions t = Base_context.user_restrictions t.base_context
+
+  let filter_deps t = Base_context.filter_deps t.base_context
+end
+
+(* Helper function for implementing [CONTEXT.filter_deps] where local package
+   dependencies include test and doc dependencies *)
+let filter_deps_env ~local_packages ~env pkg filtered_formula =
+  let pkg_is_local =
+    OpamPackage.Name.Map.mem (OpamPackage.name pkg) local_packages
+  in
+  filtered_formula
+  |> OpamFilter.partial_filter_formula (env pkg)
+  |> OpamFilter.filter_deps ~build:true ~post:true ~test:pkg_is_local
```

Comment:
It means that dependencies marked as "post" will be resolved. I've never encountered this variable in the wild but it tells opam that the dependency can be compiled after the current package. https://opam.ocaml.org/doc/Manual.html#pkgvar-post I set it to true because that's what `Opam_0install.Dir_context` does and it seems like the most useful default setting. Perhaps one day we'll need to add support to dune to set it to false to deal with dependency cycles.

Reply:
Thanks for the explanation, I have never encountered these before. Out of curiosity I checked opam-repository and there are two non-compiler packages (only 2 versions overall) that use these: `sail.0.15` and `ctypes-foreign.0.18.0` so indeed these don't seem to be used much.